### PR TITLE
RSDK-3556 - Normalize use of server and client helpers

### PR DESCRIPTION
**Pull request by stuqdog**
#### Major changes
Wherever possible, makes use of `client_helper` and `service_helper` to minimize boilerplate of resource wrapper code.

#### Minor changes
A couple small flybys (comments below highlight them).

### CODE REVIEW

File: src/viam/sdk/components/camera/client.cpp [hunk 1]
```diff
@@ -22,84 +23,44 @@ CameraClient::CameraClient(std::string name, std::shared_ptr<grpc::Channel> chan
...
-    *req.mutable_name() = this->name();
-    *req.mutable_extra() = map_to_struct(extra);
-
-    stub_->GetImage(&ctx, req, &resp);
-    return from_proto(resp);
+    return make_client_helper(this, *stub_, &StubType::GetImage)
+        .with(extra,
+              [&](auto& request) {
+                  // CR erodkin: call attention to flyby here
+                  *request.mutable_mime_type() = Camera::normalize_mime_type(mime_type);
```
Comment (stuqdog) [inline src/viam/sdk/components/camera/client.cpp:36]: (flyby) previously we weren't using the return from the `normalize_mime_type` call so the request's mime type wasn't actually being normalized.

File: src/viam/sdk/components/camera/server.cpp [hunk 2]
```diff
@@ -17,190 +18,103 @@ CameraServer::CameraServer(std::shared_ptr<ResourceManager> manager) : ResourceS
...
-    *response->mutable_mime_type() = "pointcloud/pcd";
-    *response->mutable_point_cloud() = bytes_to_string(point_cloud.pc);
-
-    return ::grpc::Status();
+    ::viam::component::camera::v1::GetPointCloudResponse* response) noexcept {
+    return make_service_helper<Camera>(
+        "CameraServer::GetPointCloud", this, request)([&](auto& helper, auto& camera) {
+        // CR erodkin: fly by here! we weren't passing extra correctly
+        const Camera::point_cloud point_cloud =
+            camera->get_point_cloud(request->mime_type(), helper.getExtra());
```
Comment (stuqdog) [inline src/viam/sdk/components/camera/server.cpp:85]: (flyby) previously, we weren't passing `extra` to this call.

File: src/viam/sdk/components/generic/server.cpp [hunk 3]
```diff
@@ -11,43 +12,25 @@ GenericServer::GenericServer(std::shared_ptr<ResourceManager> manager) : Resourc
...
-    }
-    return ::grpc::Status();
+::grpc::Status GenericServer::GetGeometries(
+    ::grpc::ServerContext* context,
+    const ::viam::common::v1::GetGeometriesRequest* request,
+    ::viam::common::v1::GetGeometriesResponse* response) noexcept {
+    return make_service_helper<Generic>(
+        "GenericServer::GetGeometries", this, request)([&](auto& helper, auto& generic) {
+        // CR erodkin: flyby adding extra
+        const std::vector<GeometryConfig> geometries = generic->get_geometries(helper.getExtra());
```
Comment (stuqdog) [inline src/viam/sdk/components/generic/server.cpp:29]: (flyby) previously, we weren't passing `extra` to this call.

File: src/viam/sdk/components/board/client.cpp [hunk 4]
```diff
@@ -23,151 +24,71 @@ BoardClient::BoardClient(std::string name, std::shared_ptr<grpc::Channel> channe
...
-        throw std::runtime_error(status.error_message());
-    }
-    return struct_to_map(response.result());
+    return make_client_helper(this, *stub_, &StubType::DoCommand)
+        .with([&](auto& request) { *request.mutable_command() = map_to_struct(command); })
+        .invoke([](auto& response) { return struct_to_map(response.result()); });
 }
 
+// TODO(RSDK-6048) update `client_wrapper` to allow for requests without a `mutable_name()` method,
+// then wrap here.
```
Comment (stuqdog) [inline src/viam/sdk/components/board/client.cpp:91]: My editor didn't catch it but this problem also prevents us from using `service_wrapper` for this method in the server code, will be backing that out as well.

File: src/viam/examples/modules/simple/main.cpp [hunk 5]
```diff
@@ -49,7 +49,7 @@ class Printer : public Generic {
         return command;
     }
 
-    std::vector<GeometryConfig> get_geometries() {
+    std::vector<GeometryConfig> get_geometries(const AttributeMap& extra) {
```
Comment (stuqdog) [inline src/viam/examples/modules/simple/main.cpp:52]: (flyby) adding `extra` support to generic

File: src/viam/sdk/components/generic/client.cpp [hunk 6]
```diff
@@ -24,10 +24,10 @@ AttributeMap GenericClient::do_command(AttributeMap command) {
...
         .invoke([](auto& response) { return struct_to_map(response.result()); });
 };
-std::vector<GeometryConfig> GenericClient::get_geometries() {
-    return make_client_helper(this, *stub_, &StubType::GetGeometries).invoke([](auto& response) {
-        return GeometryConfig::from_proto(response);
-    });
+std::vector<GeometryConfig> GenericClient::get_geometries(const AttributeMap& extra) {
+    return make_client_helper(this, *stub_, &StubType::GetGeometries)
+        .with(extra)
+        .invoke([](auto& response) { return GeometryConfig::from_proto(response); });
```
Comment (stuqdog) [inline src/viam/sdk/components/generic/client.cpp:30]: (flyby) adding `extra` support to generic

File: src/viam/sdk/components/generic/client.hpp [hunk 7]
```diff
@@ -20,7 +20,7 @@ class GenericClient : public Generic {
    public:
     GenericClient(std::string name, std::shared_ptr<grpc::Channel> channel);
     AttributeMap do_command(AttributeMap command) override;
-    std::vector<GeometryConfig> get_geometries() override;
+    std::vector<GeometryConfig> get_geometries(const AttributeMap& extra) override;
```
Comment (stuqdog) [inline src/viam/sdk/components/generic/client.hpp:23]: (flyby) adding `extra` support to generic

File: src/viam/sdk/components/generic/generic.hpp [hunk 8]
```diff
@@ -50,7 +50,11 @@ class Generic : public Component {
 
     /// @brief Returns `GeometryConfig`s associated with the calling camera.
     /// @return The requested `GeometryConfig`s associated with the component.
-    virtual std::vector<GeometryConfig> get_geometries() = 0;
+    inline std::vector<GeometryConfig> get_geometries() {
+        return get_geometries({});
+    }
+
+    virtual std::vector<GeometryConfig> get_geometries(const AttributeMap& extra) = 0;
```
Comment (stuqdog) [inline src/viam/sdk/components/generic/generic.hpp:57]: (flyby) adding `extra` support to generic

File: src/viam/sdk/tests/mocks/generic_mocks.cpp [hunk 9]
```diff
@@ -20,7 +20,7 @@ MockGeneric::do_command(
     std::shared_ptr<std::unordered_map<std::string, std::shared_ptr<ProtoType>>> command) {
     return map_;
 }
-std::vector<GeometryConfig> MockGeneric::get_geometries() {
+std::vector<GeometryConfig> MockGeneric::get_geometries(const AttributeMap& extra) {
```
Comment (stuqdog) [inline src/viam/sdk/tests/mocks/generic_mocks.cpp:23]: (flyby) adding `extra` support to generic

File: src/viam/sdk/components/board/client.cpp (hunk 4 above)
Comment (stuqdog) [reply src/viam/sdk/components/board/client.cpp:91]: Makes sense! I'll add a comment to the ticket to that effect.

**Comment by stuqdog** [comment]
> LGTM, and thanks for doing this. I didn't examine every transformation in detail, but I did skim through all of them and they all look like what I would expect to see. Did you take a pass through to make sure that all vestiges of the older style helpers that Abe and Benji had added were removed? I can't remember if I did that or not.

I _believe_ you did it? I didn't do a super thorough look through but grepped for `stub_wrapper` and `server_wrapper` (the names of the wrappers from Abe's PR) and nothing turned up.

**Comment by acmorrow** [comment]
> > LGTM, and thanks for doing this. I didn't examine every transformation in detail, but I did skim through all of them and they all look like what I would expect to see. Did you take a pass through to make sure that all vestiges of the older style helpers that Abe and Benji had added were removed? I can't remember if I did that or not.
> 
> I _believe_ you did it? I didn't do a super thorough look through but grepped for `stub_wrapper` and `server_wrapper` (the names of the wrappers from Abe's PR) and nothing turned up.

Sounds good, thanks.

**Review by benjirewis — APPROVED** [review]
Awesome! Sort of skimmed but changes LGTM, great to see so much duplicate logic get removed.

**Comment by benjirewis** [comment]
> I believe you did it? I didn't do a super thorough look through but grepped for stub_wrapper and server_wrapper (the names of the wrappers from Abe's PR) and nothing turned up.

Yep, all the templating stuff that Abe and I added seems to be gone from what I can tell.

**Pull request merged**